[FEAT] Add File to Context Automatically When No Text in Box

charlwillia6 opened the pull request:
## Description ✏️

Added automatic context pinning when the chat box is empty for the current file being edited.  It automatically switches context when switching between files.  

![image](https://github.com/user-attachments/assets/a8a34df8-f89a-4d7d-bf3b-c261c82a5017)
![image](https://github.com/user-attachments/assets/6981f5d0-6586-4b0a-a9fc-a41cccaecda3)

## Checklist ✅

- [x] I have added screenshots (if UI changes are present).
- [x] I have done a self-review of my code.
- [x] I have manually tested my code (if applicable).

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
[FEAT] Add active editor tracking and context item updates based on file changes

• Auto-add file mentions on editor change
• Improve context item cleanup on send
• Add filepath validation utilities
• Debounce editor change notifications
• Fix mention handling in TipTap editor

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main' into charlwillia6/add-file-to-context-when-no-text-selected

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
[FIX] Updated the isEmpty function to check for context length so that context does not get overwritten.

charlwillia6 committed:
[FIX] getSuggestions was causing flushSync errors, refactored to fix those errors

charlwillia6 committed:
[FIX] Issues with auto context when clicking return in the chat input box

• Move context items to component level
• Clean up mention node handling
• Improve clipboard event handling
• Optimize enter key handler
• Update context item persistence

charlwillia6 committed:
Clean-up

Fryingpannn commented on gui/src/components/mainInput/TipTapEditor.tsx:
```diff
@@ -132,7 +130,7 @@ const getPlaceholder = (historyLength: number, location: any) => {
 };
 
-function getDataUrlForFile(file: File, img): string {
+const getDataUrlForFile = (file: File, img): string => {
```
I know the urge to change the `function` to arrow functions, but let's avoid doing this x) We should minimize merge conflicts with upstream, so unfortunately it is worth the practice sacrifice.

Fryingpannn commented on gui/src/redux/slices/stateSlice.ts:
```diff
@@ -270,10 +272,9 @@ export const stateSlice = createSlice({
     },
     addContextItems: (state, action: PayloadAction<ContextItemWithId[]>) => {
-      state.contextItems = state.contextItems.concat(action.payload);
+      state.contextItems.push(...action.payload);
```
Not sure if we should use `push` here, I think for state we want things to be immutable, so should return a new object like `concat` does.

Fryingpannn committed:
Fix empty editory check - add slashcommand and codeblocks

Fryingpannn commented:
Tested. Current behaviour:

- Focused editor file will appear mentioned in chat if there's nothing in the current chat
- If there's already codeblock, slashcommand or text, then won't put current open file in chat.
- Will swap current file if no other content is in the text box upon switching files
- If close current file, the file mention still remains in chat (non-ideal)
- If have content in chat, the current mention file will not change

Fryingpannn commented on extensions/vscode/src/extension/VsCodeExtension.ts:
```diff
@@ -360,15 +361,23 @@ export class VsCodeExtension {
+
     this.ide.onDidChangeActiveTextEditor((filepath) => {
-      this.core.invoke("didChangeActiveTextEditor", { filepath });
+      this.debouncedNotifyEditorChange(filepath || null);
```
I wonder if this debouncing is needed, also with 100ms, I don't think anyone is switching that fast between tabs that it's worth it. This maybe more for typing. I'd say no need debounce here unless you feel like there are other factors that deterioriate the ux without it.

Fryingpannn committed:
Revert some refactoring

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main' into charlwillia6/add-file-to-context-when-no-text-selected

charlwillia6 committed:
Merge branch 'charlwillia6/add-file-to-context-when-no-text-selected' of https://github.com/charlwillia6/pearai-submodule into charlwillia6/add-file-to-context-when-no-text-selected

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge remote-tracking branch 'origin/main' into charlwillia6/add-file-to-context-when-no-text-selected

Himanshu-Singh-Chauhan committed:
fix: clear main input after response completes

Himanshu-Singh-Chauhan committed:
undo refactoring

Himanshu-Singh-Chauhan committed:
reverts

nang-dev2 committed:
Merge branch 'main' of https://github.com/trypear/pearai-submodule into charlwillia6/add-file-to-context-when-no-text-selected

Himanshu-Singh-Chauhan commented:
damn, this is big, took time to review.
awesome work @charlwillia6 
i tried to show "current file" text on the mention, didnt work, it does not send the file contents.
Also, I feel including current file in every chat is fine, cause user may make changes and when asks ai next step, ai can read the modified file again and see what changes were applied and what were not. 
![image](https://github.com/user-attachments/assets/53068068-4974-4b0b-be47-c83e652a7eab)

Himanshu-Singh-Chauhan commented on extensions/vscode/src/extension/VsCodeExtension.ts:
```diff
@@ -360,15 +361,23 @@ export class VsCodeExtension {
+        const filepath = event.textEditor?.document?.uri.fsPath || null;
+        this.debouncedNotifyEditorChange(filepath);
+      })
+    );
```
what is this event? is this when we select text in a file. if so, then maybe we don't want to trigger notifyeditorchange

charlwillia6 replied:
This even is specific to getting the file path of the selected file.

Himanshu-Singh-Chauhan commented on gui/src/components/mainInput/getSuggestion.ts:
```diff
@@ -21,37 +21,48 @@ function getSuggestion(
-          component = new ReactRenderer(MentionList, {
-            props: { ...props, enterSubmenu, onClose: onExit },
-            editor: props.editor,
+          requestAnimationFrame(() => {
+            component = new ReactRenderer(MentionList, {
+              props: { 
+                ...props, 
```
whats this changes, mmm, i never saw tippy tooltip, where does this show?

nang-dev replied:
Hmmm

charlwillia6 replied:
It looks like this maybe a refactoring change (not on purpose).  If you look though, it was there previously.  Scroll down and look at the diff.

nang-dev2 committed:
Merge branch 'charlwillia6/add-file-to-context-when-no-text-selected' of https://github.com/charlwillia6/pearai-submodule into charlwillia6/add-file-to-context-when-no-text-selected

nang-dev commented:
i added a pull from main . this is good so far! thx sm

2 bugs i see:
1. "New chat" should include the current file
<img width="1303" alt="Screenshot 2024-11-02 at 1 08 50 PM" src="https://github.com/user-attachments/assets/59b3b93f-b872-4be8-b576-9e61f6e1ba80">

2. All subsequent chats in any thread should have ALL of this functionality turned off / ignored.
<img width="1305" alt="Screenshot 2024-11-02 at 1 09 22 PM" src="https://github.com/user-attachments/assets/b7e9bbc8-3a3a-4a5f-a958-00200d4f266a">

nice to have: if the context used can have page.tsx (Active File)
<img width="291" alt="Screenshot 2024-11-02 at 1 07 48 PM" src="https://github.com/user-attachments/assets/279b397e-037e-4b72-9562-1797c399d65e">

charlwillia6 committed:
Move & improve file path handling in code editor for UX enhancements

• Centralized isValidFilePath utility
• Added newSession command handler
• Streamlined editor change handling
• Fixed focus management in editor

charlwillia6 commented:
@nang-dev @Fryingpannn @Himanshu-Singh-Chauhan - New changes have been pushed.  Please review and test.

Himanshu-Singh-Chauhan committed:
move newsession command to commands.ts file

Himanshu-Singh-Chauhan committed:
add currentfile on onDidChangeActiveTextEditor

Commit:
Merge remote-tracking branch 'upstream/main'

charlwillia6 committed:
Merge branch 'main' into charlwillia6/add-file-to-context-when-no-text-selected

charlwillia6 committed:
Merge remote-tracking branch 'upstream/main' into charlwillia6/add-file-to-context-when-no-text-selected

charlwillia6 commented:
This has been tested, and with @Himanshu-Singh-Chauhan changes, this seems to work just as requested.  I also merged main into this PR.

Fryingpannn committed:
Remove debounce

Fryingpannn committed:
Merge branch 'main' into charlwillia6/add-file-to-context-when-no-text-selected

Fryingpannn committed:
Add back memo

Fryingpannn commented:
Gj Charles.

Leaving some comments for future.

Notice
- If user erases context and uses their own context, then switch file, it removes the context they manually added. This is because the feature works by checking if only 1 context file, then if user switch files, also switch that context file. No distinction is made between a default active file and a file that user added, only the count. However, I think this is an obscure edge enough that we can let it slide for now. Until we have the polished active file UI in chatbox.
- for `slashCommand`s, this is currently also erased when user switches file and user only has a slash command with no text in chatbox. I think this is because it's not checking the slash command type, if already existing then should not replace it (if there's text, it won't replace). Will also let this slide for now since usually when a user uses slash command, I think not likely they will change files without even typing a single word.

Improvements ahead
- Make better UI in chatbox for "Active file" instead of just another tag in chatbox.

charlwillia6 commented:
@Fryingpannn @nang-dev @Himanshu-Singh-Chauhan - Do you want me to create an entire new PR when I fix the issues?

Second, what is your ides of this?

> Improvements ahead

>  Make better UI in chatbox for "Active file" instead of just another tag in chatbox.
